File: mobile/src/core/pairing.ts
// Copyright 2026 Geda
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Decoding the QR code a receiver shows (docs/PROTOCOL.md §3.1).
//
// This is the moment trust is established for the whole life of the pairing:
// whatever SPKI pin comes out of here is the only thing that will ever be
// accepted from that receiver again. So a payload missing any of its parts is
// rejected outright rather than half-used.

export const PAIRING_VERSION = 1;
export const PAIRING_URI_SCHEME = 'geda://pair/';

export type PairingPayload = {
  v: number;
  device_id: string;
  name: string;
  /** base64(SHA-256(SubjectPublicKeyInfo)). */
  spki: string;
  /** Every address the receiver has, `host:port`, tunnels included. */
  addrs: string[];
  /** Single-use pairing secret. */
  psk: string;
  /** Unix seconds. */
  exp: number;
};

export class PairingError extends Error {}

/** Decodes a scanned code, with or without the `geda://pair/` prefix. */
export function decodePairingPayload(scanned: string): PairingPayload {
  const trimmed = scanned.trim();
  const encoded = trimmed.startsWith(PAIRING_URI_SCHEME)
    ? trimmed.slice(PAIRING_URI_SCHEME.length)
    : trimmed;

  let parsed: unknown;
  try {
    parsed = JSON.parse(fromBase64Url(encoded));
  } catch {
    throw new PairingError('That is not a Geda Transfer pairing code.');
  }

  const payload = parsed as Partial<PairingPayload>;
  if (payload.v !== PAIRING_VERSION) {
    throw new PairingError(
      `This code is version ${String(payload.v)}; this app speaks version ${PAIRING_VERSION}. Update whichever side is older.`,
    );
  }
  if (!payload.device_id) throw new PairingError('The code names no receiver.');
  if (!payload.spki) {
    // Without the pin there is nothing to trust on first use, and the
    // connection would be an unauthenticated one wearing a QR code.
    throw new PairingError('The code carries no key to pin.');
  }
  if (!payload.psk) throw new PairingError('The code carries no pairing secret.');
  if (!payload.addrs?.length) throw new PairingError('The code carries no addresses.');

  return {
    v: payload.v,
    device_id: payload.device_id,
    name: payload.name ?? payload.device_id,
    spki: payload.spki,
    addrs: payload.addrs,
    psk: payload.psk,
    exp: payload.exp ?? 0,
  };
}

/** A pairing code is short-lived so that a photograph of a screen is worthless. */
export function isExpired(payload: PairingPayload, nowMs: number = Date.now()): boolean {
  return payload.exp !== 0 && nowMs / 1000 >= payload.exp;
}

/**
 * Turns a `host:port` candidate into a base URL.
 *
 * IPv6 addresses arrive already bracketed from the receiver; anything else is
 * left alone so that a hostname works as well as an address.
 */
export function baseUrl(addr: string): string {
  return `https://${addr}`;
}

function fromBase64Url(value: string): string {
  const padded = value.replace(/-/g, '+').replace(/_/g, '/');
  const withPadding = padded + '='.repeat((4 - (padded.length % 4)) % 4);
  return decodeBase64(withPadding);
}

/**
 * Base64 to a UTF-8 string.
 *
 * `atob` produces one character per byte, so a receiver named with anything
 * outside ASCII -- which is most of the world -- needs the bytes reassembled
 * before they are read as text.
 */
function decodeBase64(value: string): string {
  const binary = globalThis.atob(value);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) {
    bytes[i] = binary.charCodeAt(i);
  }
  return new TextDecoder('utf-8', { fatal: true }).decode(bytes);
}

File: core/pairing/pairing.go
// Copyright 2026 Geda
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package pairing carries the trust-on-first-use handshake: the QR payload a
// receiver shows, the single-use pre-shared key that authorises one pairing,
// and the per-device bearer token that results.
//
// There is no CA and no account. The user vouches for the receiver's public
// key by being physically in front of it and scanning its code; from then on
// the recorded SPKI pin is the whole of the trust relationship
// (docs/PROTOCOL.md §3).
package pairing

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"
)

// Version is the pairing payload version.
const Version = 1

// DefaultOfferTTL is how long a QR code stays valid. Long enough to walk
// across a room and unlock a phone, short enough that a photograph of a screen
// found later is worthless.
const DefaultOfferTTL = 5 * time.Minute

// URIScheme prefixes the QR payload so a camera app can route it.
const URIScheme = "geda://pair/"

// Payload is what the QR code encodes.
type Payload struct {
	V        int      `json:"v"`
	DeviceID string   `json:"device_id"`
	Name     string   `json:"name"`
	SPKI     string   `json:"spki"`
	Addrs    []string `json:"addrs"`
	PSK      string   `json:"psk"`
	Exp      int64    `json:"exp"`
}

// Encode renders the payload as base64url of compact JSON.
func (p Payload) Encode() (string, error) {
	raw, err := json.Marshal(p)
	if err != nil {
		return "", fmt.Errorf("pairing: encode payload: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(raw), nil
}

// URI renders the payload as the string to put in a QR code.
func (p Payload) URI() (string, error) {
	encoded, err := p.Encode()
	if err != nil {
		return "", err
	}
	return URIScheme + encoded, nil
}

// Expired reports whether the offer has run out.
func (p Payload) Expired(now time.Time) bool {
	return p.Exp != 0 && now.Unix() >= p.Exp
}

// ErrBadPayload reports a QR payload that cannot be used.
var ErrBadPayload = errors.New("pairing: unusable payload")

// Decode parses a scanned payload, with or without the URI scheme.
func Decode(s string) (Payload, error) {
	s = strings.TrimSpace(s)
	s = strings.TrimPrefix(s, URIScheme)

	raw, err := base64.RawURLEncoding.DecodeString(strings.TrimRight(s, "="))
	if err != nil {
		return Payload{}, fmt.Errorf("%w: %v", ErrBadPayload, err)
	}

	var p Payload
	if err := json.Unmarshal(raw, &p); err != nil {
		return Payload{}, fmt.Errorf("%w: %v", ErrBadPayload, err)
	}
	switch {
	case p.V != Version:
		return Payload{}, fmt.Errorf("%w: version %d", ErrBadPayload, p.V)
	case p.DeviceID == "":
		return Payload{}, fmt.Errorf("%w: no device id", ErrBadPayload)
	case p.SPKI == "":
		// Without the pin there is nothing to trust on first use, and the
		// connection would be an unauthenticated one wearing a QR code.
		return Payload{}, fmt.Errorf("%w: no spki pin", ErrBadPayload)
	case p.PSK == "":
		return Payload{}, fmt.Errorf("%w: no psk", ErrBadPayload)
	case len(p.Addrs) == 0:
		return Payload{}, fmt.Errorf("%w: no addresses", ErrBadPayload)
	}
	return p, nil
}

// NewPSK returns a single-use pairing secret.
func NewPSK() (string, error) { return randomToken(32) }

// NewToken returns a device bearer token.
//
// 32 bytes because the token is the only credential after pairing: it is
// checked on every request, never rotated, and revoked only by unpairing.
func NewToken() (string, error) { return randomToken(32) }

func randomToken(n int) (string, error) {
	raw := make([]byte, n)
	if _, err := rand.Read(raw); err != nil {
		return "", fmt.Errorf("pairing: generate secret: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(raw), nil
}

// Offers holds the pairing secrets a receiver is currently willing to accept.
//
// Kept in memory only: an offer that survives a restart is a credential the
// user does not know is still outstanding.
type Offers struct {
	now func() time.Time

	mu   sync.Mutex
	live map[string]time.Time // sha256(psk) -> expiry
}

// NewOffers builds an empty offer set. now may be nil.
func NewOffers(now func() time.Time) *Offers {
	if now == nil {
		now = time.Now
	}
	return &Offers{now: now, live: make(map[string]time.Time)}
}

// Issue creates an offer and returns the secret. ttl of 0 means
// DefaultOfferTTL.
func (o *Offers) Issue(ttl time.Duration) (psk string, expires time.Time, err error) {
	if ttl <= 0 {
		ttl = DefaultOfferTTL
	}

	psk, err = NewPSK()
	if err != nil {
		return "", time.Time{}, err
	}
	expires = o.now().Add(ttl)

	o.mu.Lock()
	defer o.mu.Unlock()
	o.sweepLocked()
	o.live[fingerprint(psk)] = expires

	return psk, expires, nil
}

// Redeem consumes an offer, reporting whether it was live.
//
// The lookup is by SHA-256 of the secret rather than by comparing candidates
// one at a time: hashing first means the map probe reveals nothing about the
// secret through timing, and a caller cannot walk the offer set by measuring
// how long a wrong guess takes.
func (o *Offers) Redeem(psk string) bool {
	if psk == "" {
		return false
	}
	key := fingerprint(psk)

	o.mu.Lock()
	defer o.mu.Unlock()

	expiry, ok := o.live[key]
	if !ok {
		return false
	}
	// Single use, whether or not it had expired: a secret that has been
	// presented once is spent.
	delete(o.live, key)

	return o.now().Before(expiry)
}

// Revoke drops every outstanding offer, for a user who closes the pairing
// screen.
func (o *Offers) Revoke() {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.live = make(map[string]time.Time)
}

// Len reports how many offers are live, for tests and diagnostics.
func (o *Offers) Len() int {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.sweepLocked()
	return len(o.live)
}

func (o *Offers) sweepLocked() {
	now := o.now()
	for key, expiry := range o.live {
		if !now.Before(expiry) {
			delete(o.live, key)
		}
	}
}

func fingerprint(secret string) string {
	sum := sha256.Sum256([]byte(secret))
	return hex.EncodeToString(sum[:])
}
